### PR TITLE
[RDS] API Update

### LINKED ISSUE #433: [RDS] API Updated
https://confluence.tsi-dev.otc-service.com/display/DR/RM-4222-RDS-API

### PR DESCRIPTION
**Pull request by Aloento**
fix #433

### CODE REVIEW

File: acceptance/openstack/rds/v3/flavors_test.go [hunk 1]
```diff
@@ -14,12 +14,11 @@ func TestFlavorsList(t *testing.T) {
...
-		VersionName: "10",
+		DatabaseName: "PostgreSQL",
+		VersionName:  "10",
 	}
-	allFlavorPages, err := flavors.List(client, listOpts, "PostgreSQL").AllPages()
-	th.AssertNoErr(t, err)
 
-	rdsFlavors, err := flavors.ExtractDbFlavors(allFlavorPages)
+	rdsFlavors, err := flavors.ListFlavors(client, listOpts)
 	th.AssertNoErr(t, err)
```
Comment (artem-lifshits) [inline acceptance/openstack/rds/v3/flavors_test.go:22]: flavors_test.go:22: Failure in flavors_test.go, line 22: unexpected error "json: cannot unmarshal array into Go struct field Flavor.version_name of type string"

File: acceptance/openstack/rds/v3/rds_test.go [hunk 2]
```diff
@@ -35,12 +32,12 @@ func TestRdsLifecycle(t *testing.T) {
 
 	// Create RDSv3 instance
 	rds := createRDS(t, client, cc.RegionName)
-	defer deleteRDS(t, client, rds.Id)
+	t.Cleanup(func() { deleteRDS(t, client, rds.Id) })
 	th.AssertEquals(t, rds.Volume.Size, 100)
 
-	restart, err := instances.Restart(client, instances.RestartRdsInstanceOpts{}, rds.Id).Extract()
+	restart, err := instances.Restart(client, instances.RestartOpts{InstanceId: rds.Id})
 	th.AssertNoErr(t, err)
```
Comment (artem-lifshits) [inline acceptance/openstack/rds/v3/rds_test.go]: rds_test.go:39: Failure in rds_test.go, line 39: unexpected error "Bad request with: [POST https://rds.eu-de.otc.t-systems.com/v3/a720688fb87f4575a4c000d818061eae/instances/9c497b94d2f84e839a3b18c5a01f9ee4in03/action], error message: {\"error_msg\":\"Parameter error.\",\"error_code\":\"DBS.280001\"}"

**Review by artem-lifshits — CHANGES_REQUESTED** [review]

File: openstack/rds/v3/backups/ListRestoreTimes.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+type ListRestoreTimesOpts struct {
+	// Specifies the DB instance ID.
+	InstanceId string
+	// Specifies the date to be queried. The value is in the yyyy-mm-dd format, and the time zone is UTC.
+	Date string `json:"date,omitempty"`
+}
+
+func ListRestoreTimes(client *golangsdk.ServiceClient, opts ListRestoreTimesOpts) ([]RestoreTime, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/backups/ListRestoreTimes.go:16]: No functional tests for this func

File: openstack/rds/v3/backups/RestorePITR.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	// timestamp: indicates the point-in-time restoration mode. In this mode, type is mandatory and restore_time is no mandatory.
+	Type string `json:"type" required:"true"`
+}
+
+type Target struct {
+	// Specifies the ID of the DB instance to be restored to.
+	InstanceID string `json:"instance_id" required:"true"`
+}
+
+func RestorePITR(c *golangsdk.ServiceClient, opts RestorePITROpts) (string, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/backups/RestorePITR.go:35]: No functional tests included for this

File: openstack/rds/v3/backups/ShowBackupPolicy.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package backups
+
+import (
+	golangsdk "github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+	"github.com/opentelekomcloud/gophertelekomcloud/openstack"
+)
+
+func ShowBackupPolicy(client *golangsdk.ServiceClient, instanceId string) (*BackupPolicy, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/backups/ShowBackupPolicy.go:8]: No functional tests provided for this func

File: openstack/rds/v3/backups/Update.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	// 23:00-00:00
+	StartTime string `json:"start_time,omitempty"`
+	// Specifies the backup cycle configuration. Data will be automatically backed up on the selected days every week. This parameter is mandatory except that the automated backup policy is disabled.
+	// Value range: The value is digits separated by commas (,), indicating the day of the week and starting from Monday.
+	// For example, the value 1,2,3,4 indicates that the backup period is Monday, Tuesday, Wednesday, and Thursday.
+	Period string `json:"period,omitempty"`
+}
+
+// Update accepts a UpdateOpts struct and uses the values to update a Backup.The response code from api is 200
+func Update(c *golangsdk.ServiceClient, opts UpdateOpts) (err error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/backups/Update.go:33]: No functional tests provided for this func

File: openstack/rds/v3/configurations/UpdateInstanceConfiguration.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+type UpdateInstanceConfigurationOpts struct {
+	// Specifies the DB instance ID.
+	InstanceId string
+	// Specifies the parameter values defined by users based on the default parameter templates.
+	// For example, "max_connections": "10"
+	Values map[string]interface{} `json:"values"`
+}
+
+func UpdateInstanceConfiguration(client *golangsdk.ServiceClient, opts UpdateInstanceConfigurationOpts) (*UpdateInstanceConfigurationResponse, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/configurations/UpdateInstanceConfiguration.go:17]: Functional test missing

File: openstack/rds/v3/flavors/ListDatastores.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
+package flavors
+
+import (
+	"github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+)
+
+func ListDatastores(client *golangsdk.ServiceClient, databaseName string) ([]DataStores, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/flavors/ListDatastores.go:8]: Functional test missing

File: openstack/rds/v3/flavors/ListStorageTypes.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	// Specifies the DB engine name. Its value can be any of the following and is case-insensitive:
+	// MySQL
+	// PostgreSQL
+	// SQLServer
+	DatabaseName string `json:"database_name"`
+	// Specifies the database version.
+	VersionName string `json:"version_name"`
+}
+
+func ListStorageTypes(client *golangsdk.ServiceClient, opts ListStorageTypesOpts) (*ListStorageTypesResponse, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/flavors/ListStorageTypes.go:18]: Functional test missing

File: openstack/rds/v3/instances/AttachEip.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	// NOTICE
+	// When is_bind is true, public_ip_id is mandatory.
+	// Specifies the EIP ID. The value must be in the standard UUID format.
+	PublicIpId string `json:"public_ip_id,omitempty"`
+	// true: Bind an EIP.
+	// false: Unbind an EIP.
+	IsBind bool `json:"is_bind"`
+}
+
+func AttachEip(client *golangsdk.ServiceClient, opts AttachEipOpts) (err error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/AttachEip.go:23]: functional test missing

File: openstack/rds/v3/instances/ChangeFailoverMode.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	// For MySQL, the value can be any of the following:
+	// async: asynchronous
+	// semisync: semi-synchronous
+	// For PostgreSQL, the value can be any of the following:
+	// async: asynchronous
+	// sync: synchronous
+	Mode string `json:"mode"`
+}
+
+func ChangeFailoverMode(client *golangsdk.ServiceClient, opts ChangeFailoverModeOpts) (*ChangeFailoverModeResponse, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/ChangeFailoverMode.go:21]: Functional test missing

File: openstack/rds/v3/instances/ChangeFailoverStrategy.go [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+type ChangeFailoverStrategyOpts struct {
+	InstanceId string
+	// Specifies the failover priority. Valid value:
+	// reliability: Data reliability is preferentially ensured during the failover to minimize the amount of lost data. It is recommended for services that require high data consistency.
+	// availability: Data availability is preferentially ensured during the failover to recover services quickly. It is recommended for services that have high requirements on the database online duration.
+	RepairStrategy string `json:"repairStrategy"`
+}
+
+func ChangeFailoverStrategy(client *golangsdk.ServiceClient, opts ChangeFailoverStrategyOpts) (err error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/ChangeFailoverStrategy.go:16]: Functional test missing

File: openstack/rds/v3/instances/ChangeOpsWindow.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+	// The value must be a valid value in the "HH:MM" format. The current time is in the UTC format.
+	StartTime string `json:"start_time"`
+	// Specifies the end time.
+	// The value must be a valid value in the "HH:MM" format. The current time is in the UTC format.
+	// NOTE
+	// The interval between the start time and end time must be four hours.
+	EndTime string `json:"end_time"`
+}
+
+func ChangeOpsWindow(client *golangsdk.ServiceClient, opts ChangeOpsWindowOpts) (err error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/ChangeOpsWindow.go:20]: Functional test missing

File: openstack/rds/v3/instances/ListCollations.go [hunk 14]
```diff
@@ -0,0 +1,18 @@
+package instances
+
+import (
+	golangsdk "github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+)
+
+func ListCollations(client *golangsdk.ServiceClient) ([]string, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/ListCollations.go:8]: Functional test missing

File: openstack/rds/v3/instances/MigrateFollower.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+type MigrateFollowerOpts struct {
+	InstanceId string
+	// Specifies the ID of the standby DB instance.
+	NodeId string `json:"nodeId"`
+	// Specifies the code of the AZ to which the standby DB instance is to be migrated.
+	AzCode string `json:"azCode"`
+}
+
+func MigrateFollower(client *golangsdk.ServiceClient, opts ChangeFailoverStrategyOpts) (*string, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/MigrateFollower.go]: Functional test missing

File: openstack/rds/v3/instances/SingleToHa.go [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+type SingleToHaRds struct {
+	// Specifies the AZ code of the DB instance node.
+	AzCodeNewNode string `json:"az_code_new_node" required:"true"`
+	// This parameter is mandatory only when a Microsoft SQL Server DB instance type is changed from single to primary/standby.
+	Password string `json:"password,omitempty"`
+}
+
+func SingleToHa(client *golangsdk.ServiceClient, opts SingleToHaRdsOpts) (*string, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/SingleToHa.go]: Functional test missing

File: openstack/rds/v3/instances/StartFailover.go [hunk 17]
```diff
@@ -0,0 +1,29 @@
+package instances
+
+import (
+	golangsdk "github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+)
+
+func StartFailover(client *golangsdk.ServiceClient, instanceId string) (*StartFailoverResponse, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/StartFailover.go:8]: Functional test missing

File: openstack/rds/v3/instances/StartupInstance.go [hunk 18]
```diff
@@ -0,0 +1,26 @@
+package instances
+
+import (
+	"net/http"
+
+	golangsdk "github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+)
+
+func StartupInstance(client *golangsdk.ServiceClient, instanceId string) (*string, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/StartupInstance.go:10]: Functional test missing

File: openstack/rds/v3/instances/StopInstance.go [hunk 19]
```diff
@@ -0,0 +1,13 @@
+package instances
+
+import (
+	golangsdk "github.com/opentelekomcloud/gophertelekomcloud"
+)
+
+func StopInstance(client *golangsdk.ServiceClient, instanceId string) (*string, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/StopInstance.go:7]: Functional test missing

File: openstack/rds/v3/instances/UpdateInstanceName.go [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+
+type UpdateInstanceNameOpts struct {
+	InstanceId string
+	// Specifies the DB instance name.
+	// DB instances of the same type can have same names under the same tenant.
+	// The parameter must be 4 to 64 characters long, start with a letter, and contain only letters (case-sensitive), digits, hyphens (-), and underscores (_).
+	Name string `json:"name"`
+}
+
+func UpdateInstanceName(client *golangsdk.ServiceClient, opts UpdateInstanceNameOpts) (err error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/instances/UpdateInstanceName.go:16]: Functional test missing

File: openstack/rds/v3/logs/ListErrorLog.go [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+	// LOG
+	// WARNING
+	// ERROR
+	// FATAL
+	// PANIC
+	// NOTE
+	Level string `q:"level"`
+}
+
+func ListErrorLog(client *golangsdk.ServiceClient, opts DbErrorlogOpts) (*ErrorLogResp, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/logs/ListErrorLog.go:36]: Functional test missing

File: openstack/rds/v3/logs/ListSlowLog.go [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+	//
+	// INSERT
+	// UPDATE
+	// SELECT
+	// DELETE
+	// CREATE
+	Level string `q:"level"`
+}
+
+func ListSlowLog(client *golangsdk.ServiceClient, opts DbSlowLogOpts) (*SlowLogResp, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/logs/ListSlowLog.go:32]: Functional test missing

File: openstack/rds/v3/security/SetSecurityGroup.go [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/build"
+)
+
+type SetSecurityGroupOpts struct {
+	InstanceId string
+	// Specifies the security group ID.
+	SecurityGroupId string `json:"security_group_id"`
+}
+
+func SetSecurityGroup(c *golangsdk.ServiceClient, opts UpdateDataIpOpts) (*string, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/security/SetSecurityGroup.go]: Functional test missing

File: openstack/rds/v3/security/SwitchSsl.go [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+type SwitchSslOpts struct {
+	// Specifies the DB instance ID.
+	InstanceId string
+	// Specifies whether to enable SSL.
+	// true: SSL is enabled.
+	// false: SSL is disabled.
+	SslOption bool `json:"ssl_option"`
+}
+
+func SwitchSsl(c *golangsdk.ServiceClient, opts SwitchSslOpts) (err error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/security/SwitchSsl.go:18]: Functional test missing

File: openstack/rds/v3/security/UpdateDataIp.go [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/build"
+)
+
+type UpdateDataIpOpts struct {
+	InstanceId string
+	// Indicates the private IP address.
+	NewIp string `json:"new_ip"`
+}
+
+func UpdateDataIp(c *golangsdk.ServiceClient, opts UpdateDataIpOpts) (*string, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/security/UpdateDataIp.go:14]: Functional test missing

File: openstack/rds/v3/security/UpdatePort.go [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+	// The PostgreSQL database port ranges from 2100 to 9500.
+	// The Microsoft SQL Server database port is 1433 or ranges from 2100 to 9500 (excluding 5355 and 5985).
+	// The default values is as follows:
+	// The default value of MySQL is 3306.
+	// The default value of PostgreSQL is 5432.
+	// The default value of Microsoft SQL Server is 1433.
+	Port int32 `json:"port"`
+}
+
+func UpdatePort(c *golangsdk.ServiceClient, opts SwitchSslOpts) (*string, error) {
```
Comment (artem-lifshits) [inline openstack/rds/v3/security/UpdatePort.go]: Functional test missing

**Review by artem-lifshits — CHANGES_REQUESTED** [review]

File: acceptance/openstack/rds/v3/rds_test.go [hunk 27]
```diff
@@ -56,89 +45,335 @@ func TestRdsLifecycle(t *testing.T) {
...
+		t.Logf("RDSv3 Read Replica instance deleted: %s", replica.Instance.Id)
+	})
 
-	createResult := instances.CreateReplica(client, createOpts)
-	rdsReadReplica, err := createResult.Extract()
+	t.Log("UpdateDataIp")
+
+	ip, err := security.UpdateDataIp(client, security.UpdateDataIpOpts{
+		InstanceId: rds.Id,
+		NewIp:      "192.168.30.254",
```
Comment (anton-sidelnikov) [inline acceptance/openstack/rds/v3/rds_test.go:193]: it possible to obtain this ip from subnet cidr + random?

File: acceptance/openstack/rds/v3/rds_test.go [hunk 28]
```diff
@@ -56,89 +45,335 @@ func TestRdsLifecycle(t *testing.T) {
...
+	// }
+	// TODO: Action Forbidden
+	// follower, err := instances.MigrateFollower(client, instances.MigrateFollowerOpts{
+	// 	InstanceId: rds.Id,
+	// 	NodeId:     haRds.Instances[0].Nodes[0].Id,
+	// 	AzCode:     az2,
+	// })
+	// th.AssertNoErr(t, err)
+	// err = instances.WaitForJobCompleted(client, 600, *follower)
+	// th.AssertNoErr(t, err)
```
Comment (anton-sidelnikov) [inline acceptance/openstack/rds/v3/rds_test.go]: if this api not working please add some description about what returns

File: acceptance/openstack/rds/v3/rds_test.go [hunk 29]
```diff
@@ -2,46 +2,35 @@ package v3
...
-	rdsInstances, err := instances.ExtractRdsInstances(allRdsPages)
-	th.AssertNoErr(t, err)
-
-	for _, rds := range rdsInstances.Instances {
-		tools.PrintResource(t, rds)
-	}
-}
-
 func TestRdsLifecycle(t *testing.T) {
 	client, err := clients.NewRdsV3()
```
Comment (anton-sidelnikov) [inline acceptance/openstack/rds/v3/rds_test.go:26]: skip this test from running in zuul, too long to complete

File: openstack/rds/v3/backups/testing/requests_test.go [hunk 30]
```diff
@@ -12,51 +12,6 @@ import (
 	"github.com/opentelekomcloud/gophertelekomcloud/testhelper/client"
```
Comment (anton-sidelnikov) [inline openstack/rds/v3/backups/testing/requests_test.go]: it is possible to move this into acceptance?

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]

File: acceptance/openstack/rds/v3/rds_test.go (hunk 29 above)
Comment (anton-sidelnikov) [reply acceptance/openstack/rds/v3/rds_test.go:26]: if os.Getenv("RUN_RDS_LIFECYCLE") == "" {
		t.Skip("too slow to run in zuul")
	}

File: acceptance/openstack/rds/v3/rds_test.go (hunk 27 above)
Comment (Aloento) [reply acceptance/openstack/rds/v3/rds_test.go:193]: It's basically a random private IP, so it's not very important, and we don't have a list of available private IPs (only a list of used) API.

File: openstack/rds/v3/backups/testing/requests_test.go (hunk 30 above)
Comment (Aloento) [reply openstack/rds/v3/backups/testing/requests_test.go]: done

File: acceptance/openstack/rds/v3/rds_test.go (hunk 29 above)
Comment (Aloento) [reply acceptance/openstack/rds/v3/rds_test.go:26]: done

File: acceptance/openstack/rds/v3/rds_test.go (hunk 28 above)
Comment (Aloento) [reply acceptance/openstack/rds/v3/rds_test.go]: done

**Review by anton-sidelnikov — DISMISSED** [review]

**Review by anton-sidelnikov — APPROVED** [review]

**Review by artem-lifshits — APPROVED** [review]

**Comment by Aloento** [comment]
regate

**Comment by Aloento** [comment]
regate

**Pull request merged**